src/protocol/codecs/payload.rs

```diff
@@ -0,0 +1,65 @@
+use std::io;
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+use tracing::Span;
+
+use crate::protocol::{
+    codecs::{
+        tagmsg::Tag,
+        topic::{MsgOfInterest, TopicCodec},
+    },
+    invalid_data,
+};
+
+#[derive(Debug)]
+pub enum Payload {
+    MsgOfInterest(MsgOfInterest),
+    NotImplemented,
+}
+
+/// [PayloadCodec] decodes the Algod message payload using a provided tag.
+#[derive(Clone)]
+pub struct PayloadCodec {
+    /// The associated node's span.
+    span: Span,
+
+    /// Represents a message payload type identifier.
+    // Should be set by the outer codec so that this codec knows how to interpret the payload.
+    pub tag: Option<Tag>,
+
+    /// Codec for topics which are key-value string pairs.
+    topic: TopicCodec,
+}
+
+impl PayloadCodec {
+    pub fn new(span: Span) -> Self {
+        Self {
+            span,
+            tag: None,
+            topic: TopicCodec::default(),
```

Comment:
```suggestion
            ..Default::default()
```
can be done in a follow-up :smiley: